### ISSUE TITLE
never type does not work as expected in the function of a class

### ISSUE THREAD

**Issue by SkyGrasperKnight** [microsoft/TypeScript#54265]
# Bug Report

### 🔎 Search Terms

- never
- class
- function

### 🕗 Version & Regression Information

- Version: 5.0.4
- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `never` type and class

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=5.0.4#code/GYVwdgxgLglg9mABFAFgJzgdwKJo2gMXGnjAAoBKRALkTAFMA3etRAbwChFl0s77MiXPjIAiYXFYwkoSLASiKHAL4cOEADYBDAM46heSQGUWjGBHrsuPDDkNoAwtr2VaDZq07duqW-0ESaGKBiNKImro6itaqqmqyJAj8HgBycFAA6pIA1tIA5gCCOtgAHgAO9ND0ACaUVtwQCDpQiCz4JmhmFogAvP4G7abm9JQA3NZtxkMWAHS+WIFOkWPW1gD0a4gAKiiWwHAaGlj5iBrSljoocCAa1YgARpbV9FCVr3e6iOBo9FoQKFp7hp6NZ5nZ8EQ5KQVnEEvIkO4WFk0LkwIViuU3jU6l4bAt7JDEuQKON1psdnsDkdMCdGs9Qvpnq8qh99N9fv9AcDrI0wM1WvYOl1LH0GAFBdMRiSJhLOsM5rxwZIli5pcogA)

### 💻 Code

```ts
function throwErrorFunction() : never {
  throw new Error("Error in function")
}

class ErrorService {
  throwErrorClass(): never {
    throw new Error("Error in class")
  }
}


function neverNotWorkingAsExpected() {
  const errorService = new ErrorService();
  errorService.throwErrorClass();
  
  // The following line should be detected as unreachable
  throwErrorFunction();
}

function neverWorkingAsExpected() {
  throwErrorFunction();

  // The following code is detected as unreachable
  const errorService = new ErrorService();
  errorService.throwErrorClass();
}
```

### 🙁 Actual behavior

Code following a function of a class of type 'never' is not considered inaccessible.
It does not have the same behavior as a function that is not in a class.

### 🙂 Expected behavior

The `never` type is correctly handled when used in a function of a class.
So after using the function, the following code is considered unreachable.

**Comment by MartinJohns** [comment]
It's a known design limitation, see #32695.

> A function call is analyzed as an assertion call or never-returning call when
> 
> - the call occurs as a top-level expression statement, and
> - the call specifies a single identifier or a dotted sequence of identifiers for the function name, and
> - each identifier in the function name references an entity with an explicit type, and
> - the function name resolves to a function type with an `asserts` return type or an explicit `never` return type annotation.


Duplicate of #36753.

**Comment by fatcerberus** [comment]
In particular, the important part here is...

> each identifier in the function name references an entity with an explicit type

"Function name" meaning "call target", in this case `errorService.throwErrorClass`.  `errorService` doesn't have a type annotation.
